### PR TITLE
Fix: Off-chain and on-chain data inconsistency

### PR DESCRIPTION
**Pull request by josipbagaric**
**Previous workflow:**
1. Calculate off-chain data hash
2. Store hash on-chain
3. Store off-chain data to the backend

**Current improved workflow:**
1. Calculate off-chain data hash
2. Store off-chain data to the backend
3. Store the hash on-chain
4. If on-chain data failed to be written, delete off-chain data

### CODE REVIEW

File: packages/market/src/blockchain-facade/Supply.ts [hunk 1]
```diff
@@ -92,12 +92,21 @@ export const createSupply = async (
...
 
-    if (configuration.offChainDataSource) {
-        url = supply.getUrl();
-        propertiesDocumentHash = offChainStorageProperties.rootHash;
-    }
+    url = supply.getUrl();
+    propertiesDocumentHash = offChainStorageProperties.rootHash;
+
+    let idExists = false;
+    do {
```
Comment (kosecki123) [inline packages/market/src/blockchain-facade/Supply.ts]: Such a construct has a potential to put the system on freeze, infinite loop when for some reason offchain service is down. 

The recommended way is to use tools like poly-js with exponential time ramp.

**Comment by kosecki123** [comment]
@JosephBagaric what about using polly on origin-backend-client for `get`?

**Review by e00dan — APPROVED** [review]

**Review by e00dan — CHANGES_REQUESTED** [review]
Demo deployment fails:

```
➜  origin git:(off/on-chain-data-sync) yarn run:demo
yarn run v1.19.1
$ lerna run start --scope @energyweb/utils-demo --stream
lerna notice cli v3.18.4
lerna info versioning independent
lerna notice filter including "@energyweb/utils-demo"
lerna info filter [ '@energyweb/utils-demo' ]
lerna info Executing command in 1 package: "yarn run start"
@energyweb/utils-demo: $ ts-node src/deploy
@energyweb/utils-demo: -----------------------------------------------------------
@energyweb/utils-demo: info: UserLogic Contract Deployed: 0x8a9C77e77CA52916e6fec0AB70bC9833309237D9
@energyweb/utils-demo: info: AssetLogic Contract Deployed: 0xC54F5555B81Ce48CCfb5b8CbA5F9E3Ac9790EecC
@energyweb/utils-demo: info: CertificateLogic Contract Deployed: 0xbc8c2366dF8655623384672e5c58081424359548
@energyweb/utils-demo: info: MarketLogic Deployed: 0x039f7C7c5aEd6E674d31813b2299052aB65bCe2e
@energyweb/utils-demo: -----------------------------------------------------------
@energyweb/utils-demo: verbose: http://localhost:3030/api/User/0x8a9c77e77ca52916e6fec0ab70bc9833309237d9 json is valid
@energyweb/utils-demo: verbose: Put off chain properties to http://localhost:3030/api/User/0x8a9c77e77ca52916e6fec0ab70bc9833309237d9/0xD173313A51f8fc37BcF67569b463abd89d81844f
@energyweb/utils-demo: (node:20146) UnhandledPromiseRejectionWarning: Error: createAsset: Saving off-chain data failed.
@energyweb/utils-demo:     at Entity.syncOffChainStorage (/home/kuzi/Projects/origin/packages/utils-general/src/blockchain-facade/BlockchainDataModelEntity.ts:72:19)
@energyweb/utils-demo:     at processTicksAndRejections (internal/process/task_queues.js:93:5)
@energyweb/utils-demo:     at Object.exports.createMarketUser (/home/kuzi/Projects/origin/packages/market/src/blockchain-facade/MarketUser.ts:65:5)
@energyweb/utils-demo:     at Object.exports.marketDemo (/home/kuzi/Projects/origin/packages/utils-demo/src/market.ts:89:5)
@energyweb/utils-demo:     at main (/home/kuzi/Projects/origin/packages/utils-demo/src/deploy.ts:14:5)
@energyweb/utils-demo: (node:20146) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
@energyweb/utils-demo: (node:20146) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
lerna success run Ran npm script 'start' in 1 package in 11.2s:
lerna success - @energyweb/utils-demo
Done in 11.61s.
➜  origin git:(off/on-chain-data-sync) 
```

**Comment by josipbagaric** [comment]
> Demo deployment fails:
> 
> ```
> ➜  origin git:(off/on-chain-data-sync) yarn run:demo
> yarn run v1.19.1
> $ lerna run start --scope @energyweb/utils-demo --stream
> lerna notice cli v3.18.4
> lerna info versioning independent
> lerna notice filter including "@energyweb/utils-demo"
> lerna info filter [ '@energyweb/utils-demo' ]
> lerna info Executing command in 1 package: "yarn run start"
> @energyweb/utils-demo: $ ts-node src/deploy
> @energyweb/utils-demo: -----------------------------------------------------------
> @energyweb/utils-demo: info: UserLogic Contract Deployed: 0x8a9C77e77CA52916e6fec0AB70bC9833309237D9
> @energyweb/utils-demo: info: AssetLogic Contract Deployed: 0xC54F5555B81Ce48CCfb5b8CbA5F9E3Ac9790EecC
> @energyweb/utils-demo: info: CertificateLogic Contract Deployed: 0xbc8c2366dF8655623384672e5c58081424359548
> @energyweb/utils-demo: info: MarketLogic Deployed: 0x039f7C7c5aEd6E674d31813b2299052aB65bCe2e
> @energyweb/utils-demo: -----------------------------------------------------------
> @energyweb/utils-demo: verbose: http://localhost:3030/api/User/0x8a9c77e77ca52916e6fec0ab70bc9833309237d9 json is valid
> @energyweb/utils-demo: verbose: Put off chain properties to http://localhost:3030/api/User/0x8a9c77e77ca52916e6fec0ab70bc9833309237d9/0xD173313A51f8fc37BcF67569b463abd89d81844f
> @energyweb/utils-demo: (node:20146) UnhandledPromiseRejectionWarning: Error: createAsset: Saving off-chain data failed.
> @energyweb/utils-demo:     at Entity.syncOffChainStorage (/home/kuzi/Projects/origin/packages/utils-general/src/blockchain-facade/BlockchainDataModelEntity.ts:72:19)
> @energyweb/utils-demo:     at processTicksAndRejections (internal/process/task_queues.js:93:5)
> @energyweb/utils-demo:     at Object.exports.createMarketUser (/home/kuzi/Projects/origin/packages/market/src/blockchain-facade/MarketUser.ts:65:5)
> @energyweb/utils-demo:     at Object.exports.marketDemo (/home/kuzi/Projects/origin/packages/utils-demo/src/market.ts:89:5)
> @energyweb/utils-demo:     at main (/home/kuzi/Projects/origin/packages/utils-demo/src/deploy.ts:14:5)
> @energyweb/utils-demo: (node:20146) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
> @energyweb/utils-demo: (node:20146) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
> lerna success run Ran npm script 'start' in 1 package in 11.2s:
> lerna success - @energyweb/utils-demo
> Done in 11.61s.
> ➜  origin git:(off/on-chain-data-sync) 
> ```

Thanks testing the deployment. We should add this check to the build process.

**Review by e00dan — APPROVED** [review]
👍

**Review by kosecki123 — APPROVED** [review]

**Comment by kosecki123** [comment]
> 👍

☝️

**Pull request merged**